### PR TITLE
Noise: support for leakage in MPS 

### PR DESCRIPTION
**Pull request by murogrande**
Related to #161, 
The leakage will be our 3rd, basis $|0\rangle = [0,0,1]$. In this MR, I am modifying the `mps.py` file and also the `test_mps.py` in order to support 3 level systems. 
- The new basis is: ${"r","g", "x"}$.
- When the leakage is activated, the `sample` will threat this as a "0". I add tests in `test_probablitysample.py` where you will see this behavior. 

NOTE: the correlation matrix and the occupation operators are not supported yet. Thus, in `test_mps.py` I am not modifying these. I will do it in the next MR where I will update the MPO class in order to support 3 level systems

### CODE REVIEW

File: emu_mps/mps.py [hunk 1]
```diff
@@ -104,11 +110,22 @@ def make(
...
+
+        elif len(eigenstates) == 3:  # (g,r,x)
+            ground_state = [
+                torch.tensor([[[1.0], [0.0], [0.0]]], dtype=dtype)
                 for _ in range(num_sites)
-            ],
+            ]
+
+        else:
+            raise ValueError("Unsupported qudit provided. Upto 3 levels supported.")
```
Comment (a-quelle-pasqal) [inline emu_mps/mps.py]: what about digital, that has something like g and h no?

File: test/emu_mps/test_mps.py [hunk 2]
```diff
@@ -9,6 +9,15 @@
 
 from test.utils_testing import ghz_state_factors
 
+dtype = torch.complex128
+device = "cpu"
```
Comment (a-quelle-pasqal) [inline test/emu_mps/test_mps.py]: is this used?

File: emu_mps/mps.py [hunk 3]
```diff
@@ -67,6 +71,7 @@ def __init__(
 
         self.factors = factors
         self.num_sites = len(factors)
+        self.dim = len(self.eigenstates)
```
Comment (a-quelle-pasqal) [inline emu_mps/mps.py]: I think we should assert that the physical dimension of all the state tensors is equal to this. When you add this, you will notice that some tests start failing.

File: emu_mps/mps.py (hunk 1 above)
Comment (murogrande) [reply emu_mps/mps.py]: but we are not supporting the `digital` basis in emulators but I see your point. Let me specify this:
`Unsupported basis provided. The supported basis are: `{(0,1),("r","g"),("r","g","x")} `
What do you think about this comment?

File: test/emu_mps/test_endtoend.py [hunk 4]
```diff
@@ -328,8 +328,8 @@ def test_end_to_end_afm_ring() -> None:
     assert max_bond_dim == 29
     assert fidelity_st.overlap(state_fin) == approx(fidelity_fin, abs=1e-10)
 
-    assert bitstrings["1010101010"] == 129
-    assert bitstrings["0101010101"] == 135
+    assert bitstrings["1010101010"] == 138
```
[the remark this answers is not in the document]
Comment (murogrande) [reply test/emu_mps/test_endtoend.py]: I update now all the failing tests are patched with the `cpu_multinomial_wrapper`

File: emu_mps/mps.py (hunk 1 above)
Comment (a-quelle-pasqal) [reply emu_mps/mps.py]: yes, that's better

File: emu_mps/custom_callback_implementations.py [hunk 5]
```diff
@@ -25,9 +25,18 @@ def qubit_occupation_mps_impl(
...
-    op = torch.tensor(
-        [[[0.0, 0.0], [0.0, 1.0]]], dtype=torch.complex128, device=state.factors[0].device
-    )
+    if state.dim == 2:
+        op = torch.tensor(
+            [[[0.0, 0.0], [0.0, 1.0]]],
+            dtype=torch.complex128,
+            device=state.factors[0].device,
+        )
+    elif state.dim == 3:
```
[the remark this answers is not in the document]
Comment (a-quelle-pasqal) [reply emu_mps/custom_callback_implementations.py]: that is a good point, you could create a `torch.zeros(state.dim,state.dim)` and then just put the `1,1` element to `1` it would collapse all the if-else logic into those two lines.

File: emu_mps/mps.py [hunk 6]
```diff
@@ -182,7 +201,7 @@ def get_max_bond_dim(self) -> int:
         Returns:
             the largest bond dimension in the state
         """
-        return max((x.shape[2] for x in self.factors), default=0)
+        return max((factor.shape[2] for factor in self.factors), default=0)
```
[the remark this answers is not in the document]
Comment (a-quelle-pasqal) [reply emu_mps/mps.py]: I've also seen situations where the bond-dimension oscillates a bit, so on average it grows towards the inside, but then the odd sites have slightly higher entanglement than the even or vice versa. So in general, there is no way of knowing at what site the bond dimension peaks.

File: emu_mps/mps.py [hunk 7]
```diff
@@ -40,18 +41,21 @@ def __init__(
         eigenstates: Sequence[Eigenstate] = ("r", "g"),
     ):
         """
-        This constructor creates a MPS directly from a list of tensors. It is for internal use only.
+        This constructor creates a MPS directly from a list of tensors. It i
```
[the remark this answers is not in the document]
Comment (murogrande) [reply emu_mps/mps.py]: good Elie !

File: emu_base/pulser_adapter.py [hunk 8]
```diff
@@ -84,13 +84,16 @@ def _extract_omega_delta_phi(
     elif "XY" in sequence_dict and len(sequence_dict) == 1:
         locals_a_d_p = sequence_dict["XY"]
     else:
-        raise ValueError("Only `ground-rydberg` and `mw_global` channels are supported.")
+        raise ValueError(
+            "Only `ground-rydberg` and `mw_global` channels " "are supported."
```
Comment (kbidzhiev) [inline emu_base/pulser_adapter.py]: channels " "are -> channels are

File: test/emu_mps/test_probabilitysample.py [hunk 9]
```diff
@@ -1,30 +1,49 @@
...
+
 dtype = torch.complex128
 
 
-def test_sampling_ghz5_mps():
-    device = "cpu"
+@pytest.mark.parametrize(
+    "basis",
+    (
+        ("0", "1"),
```
Comment (kbidzhiev) [inline test/emu_mps/test_probabilitysample.py]: just (r, g) as well to check the full space ? though the sampling part is expected do to the same

File: test/emu_mps/test_probabilitysample.py [hunk 10]
```diff
@@ -37,7 +56,94 @@ def test_not_orthogonalized_state():
...
+
+
+def test_with_leakage():
+    torch.manual_seed(seed)
+    shots = 1000
+    half_atoms = 4  # half the number of qubits
+
+    ket1 = torch.tensor([[[0], [1], [0]]], dtype=torch.complex128, device=device)  # |r>
+    ket0 = torch.tensor([[[1], [0], [0]]], dtype=torch.complex128, device=device)  # |g>
+    wall_state = [ket1] * half_atoms + [ket0] * half_atoms  # wall state in rydberg basis
```
Comment (kbidzhiev) [inline test/emu_mps/test_probabilitysample.py]: correct name is "domain wall"

File: test/utils_testing/utils_testing.py [hunk 11]
```diff
@@ -6,44 +6,120 @@
 import pulser
 import math
 
+# for testing purposes, reference to the real multinomial
+_real_multinomial = torch.multinomial
+
+
+def cpu_multinomial_wrapper(probs: torch.Tensor, num_samples: int, replacement=False):
+    """
+    For independent device *(cpu or gpu) tests. This is a function that
```
Comment (kbidzhiev) [inline test/utils_testing/utils_testing.py]: what * means ?

File: test/utils_testing/utils_testing.py [hunk 12]
```diff
@@ -6,44 +6,120 @@
...
-        / torch.sqrt(torch.tensor([2.0], device=device, dtype=dtype))
-        * torch.tensor(
+    if dim == 2:
+        core_1 = (
+            1
+            / torch.sqrt(torch.tensor([2.0], device=device, dtype=dtype))
+            * torch.tensor(
+                [
+                    [
+                        [1.0 + 0.0j, 0.0 + 0.0j],
```
Comment (kbidzhiev) [inline test/utils_testing/utils_testing.py]: you can safely remove `0.0j` and make 1.0 to 1, because type you use is complex

File: test/emu_mps/test_probabilitysample.py (hunk 10 above)
Comment (murogrande) [reply test/emu_mps/test_probabilitysample.py]: Thanks for the clarification

File: test/utils_testing/utils_testing.py (hunk 11 above)
Comment (murogrande) [reply test/utils_testing/utils_testing.py]: the * is an error. Good catch

File: emu_mps/mps.py [hunk 13]
```diff
@@ -206,64 +240,52 @@ def sample(
...
         while shots_done < num_shots:
             batch_size = min(max_batch_size, num_shots - shots_done)
             batched_accumulator = torch.ones(
-                batch_size, 1, dtype=torch.complex128, device=self.factors[0].device
+                batch_size, 1, dtype=dtype, device=self.factors[0].device
             )
 
-            batch_outcomes = torch.empty(batch_size, self.num_sites, dtype=torch.bool)
-
+            batch_outcomes = torch.empty(batch_size, self.num_sites, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (kbidzhiev) [reply emu_mps/mps.py]: this is a datatype, I don't see any reason why it can cause runtime error. You might say that the number of samples can be large, but default `int` is large enough. I don't have anything against `long`. It just looks out of turn and unnecessary

[the remark this answers is not in the document]
Comment (murogrande) [reply emu_mps/mps.py]: I will change it

File: test/emu_mps/test_probabilitysample.py [hunk 14]
```diff
@@ -37,7 +57,96 @@ def test_not_orthogonalized_state():
...
+    assert bitstrings.get("000") == 501
+
+
+def test_sampling_wall():
+    torch.manual_seed(seed)
+    shots = 1000
+    half_atoms = 4  # half the number of qubits
+
+    factor1 = torch.tensor([[[0], [1]]], dtype=torch.complex128, device=device)
+    factor2 = torch.tensor([[[1.0], [0.0]]], dtype=torch.complex128, device=device)
```
Comment (kbidzhiev) [inline test/emu_mps/test_probabilitysample.py]: the same 1.0 can be 1 and 0.0 can be 0

File: test/utils_testing/utils_testing.py [hunk 15]
```diff
@@ -6,44 +6,120 @@
...
-        / torch.sqrt(torch.tensor([2.0], device=device, dtype=dtype))
-        * torch.tensor(
+    if dim == 2:
+        core_1 = (
+            1
+            / torch.sqrt(torch.tensor([2.0], device=device, dtype=dtype))
+            * torch.tensor(
+                [
+                    [
+                        [1.0, 0.0],
```
[the remark this answers is not in the document]
Comment (kbidzhiev) [reply test/utils_testing/utils_testing.py]: okaaay :)
